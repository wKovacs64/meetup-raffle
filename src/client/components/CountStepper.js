import React from 'react';
import PropTypes from 'prop-types';
import useStepper from '../use-stepper';

const CountStepper = ({
  inputId,
  labelText,
  min,
  max,
  field,
  form: { setFieldValue },
}) => {
  const {
    getInputProps,
    getIncrementProps,
    getDecrementProps,
    value,
  } = useStepper({
    min,
    max,
    enableReinitialize: true,
    defaultValue: parseInt(field.value, 10) || 1,
    onNewValue: newValue => setFieldValue(field.name, newValue),
  });

  return (
    <React.Fragment>
      {labelText && (
        <div className="mb3">
          <label className="f4 f3-ns lh-copy dark-blue" htmlFor={inputId}>
            {labelText}
          </label>
        </div>
      )}
      <span className="flex flex-row">
        <button
          aria-label="decrement"
          type="button"
          disabled={value <= min}
          className={`bn bg-transparent h3 w3 pointer ${
            value <= min ? 'silver' : 'near-black'
          }`}
          data-testid="decrement-button"
          {...getDecrementProps()}
        >
          <svg
            xmlns="http://www.w3.org/2000/svg"
            fill="currentColor"
            viewBox="0 0 40 40"
            preserveAspectRatio="xMidYMid meet"
          >
            <g>
              <path d="m22.5 17.5v-5h-5v5h-5l7.5 10 7.5-10h-5z" />
            </g>
          </svg>
        </button>
        <input
          id={inputId}
          className="tc near-black w3 f3 pv1"
          pattern="[0-9]*"
          {...getInputProps({ ...field })}
        />
        <button
          aria-label="increment"
          type="button"
          disabled={value >= max}
          className={`bn bg-transparent h3 w3 pointer ${
            value >= max ? 'silver' : 'near-black'
          }`}
          data-testid="increment-button"
          {...getIncrementProps()}
        >
          <svg
            xmlns="http://www.w3.org/2000/svg"
            fill="currentColor"
            viewBox="0 0 40 40"
            preserveAspectRatio="xMidYMid meet"
          >
            <g>
              <path d="m20 12.5l-7.5 10h5v5h5v-5h5l-7.5-10z" />
            </g>
          </svg>
        </button>
      </span>
    </React.Fragment>
  );
};

CountStepper.propTypes = {
  inputId: PropTypes.string.isRequired,
  labelText: PropTypes.string,
  min: PropTypes.number,
  max: PropTypes.number,
  field: PropTypes.shape({
    value: PropTypes.any.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  form: PropTypes.shape({
    setFieldValue: PropTypes.func.isRequired,
  }).isRequired,
};

CountStepper.defaultProps = {
  labelText: '',
  min: 1,
  max: 9,
};

CountStepper.displayName = 'CountStepper';

export default CountStepper;
