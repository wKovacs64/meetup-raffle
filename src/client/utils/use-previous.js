import React from 'react';

export const usePrevious = (value) => {
  const previousValue = React.useRef();
  React.useEffect(() => {
    previousValue.current = value;
  }, [value]);
  return previousValue.current;
};
