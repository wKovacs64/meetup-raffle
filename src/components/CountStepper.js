import React from 'react';
import PropTypes from 'prop-types';
import Stepper from 'react-stepper-primitive';

const CountStepper = ({ inputId, field, form: { setFieldValue } }) => (
  <Stepper
    min={1}
    defaultValue={parseInt(field.value, 10)}
    value={parseInt(field.value, 10)}
    onChange={value => setFieldValue(field.name, value)}
    render={({ getInputProps, getIncrementProps, getDecrementProps }) => (
      <span className="flex flex-row">
        <button
          aria-label="decrement"
          type="button"
          className="bn bg-transparent near-black h3 w3 pointer"
          {...getDecrementProps()}
        >
          <svg
            xmlns="http://www.w3.org/2000/svg"
            fill="currentColor"
            viewBox="0 0 40 40"
            preserveAspectRatio="xMidYMid meet"
          >
            <g>
              <path d="m22.5 17.5v-5h-5v5h-5l7.5 10 7.5-10h-5z" />
            </g>
          </svg>
        </button>
        <input
          id={inputId}
          className="tc near-black w3 f3 pv1"
          {...field}
          {...getInputProps()}
        />
        <button
          aria-label="increment"
          type="button"
          className="bn bg-transparent near-black h3 w3 pointer"
          {...getIncrementProps()}
        >
          <svg
            xmlns="http://www.w3.org/2000/svg"
            fill="currentColor"
            viewBox="0 0 40 40"
            preserveAspectRatio="xMidYMid meet"
          >
            <g>
              <path d="m20 12.5l-7.5 10h5v5h5v-5h5l-7.5-10z" />
            </g>
          </svg>
        </button>
      </span>
    )}
  />
);

CountStepper.propTypes = {
  inputId: PropTypes.string.isRequired,
  field: PropTypes.shape({
    value: PropTypes.any.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  form: PropTypes.shape({
    setFieldValue: PropTypes.func.isRequired,
  }).isRequired,
};

export default CountStepper;
