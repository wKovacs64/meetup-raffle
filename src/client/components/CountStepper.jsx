import * as React from 'react';
import PropTypes from 'prop-types';
import useStepper from 'use-stepper';
import { usePrevious } from '../utils';

function CountStepper({
  inputId,
  labelText,
  min,
  max,
  defaultValue,
  onNewValue,
}) {
  function validValueClosestTo(newValue) {
    return String(Math.min(max, Math.max(newValue, min)));
  }

  function countReducer(state, action) {
    const integerValue = parseInt(state.value, 10);
    switch (action.type) {
      case useStepper.actionTypes.increment: {
        const newValue = validValueClosestTo(integerValue + 1);
        /* c8 ignore else */
        if (newValue !== state.value) {
          return { value: newValue };
        }
        return state;
      }
      case useStepper.actionTypes.decrement: {
        const newValue = validValueClosestTo(integerValue - 1);
        /* c8 ignore else */
        if (newValue !== state.value) {
          return { value: newValue };
        }
        return state;
      }
      case useStepper.actionTypes.coerce: {
        if (Number.isNaN(integerValue)) {
          return { value: String(defaultValue) };
        }
        /* c8 ignore else */
        if (integerValue !== state.value) {
          return { value: validValueClosestTo(integerValue) };
        }
        return state;
      }
      default:
        return useStepper.defaultReducer(state, action);
    }
  }

  const { getInputProps, getIncrementProps, getDecrementProps, value } =
    useStepper({
      min,
      max,
      defaultValue,
      enableReinitialize: true,
      stateReducer: countReducer,
    });

  const previousValue = usePrevious(value);

  React.useEffect(() => {
    // only call back with a new value if there was a previous value to avoid
    // sending the initial value change to the raffle machine before it's ready
    if (previousValue !== undefined) onNewValue(value);
  }, [onNewValue, previousValue, value]);

  const numericValue = parseFloat(value);
  const decDisabled = numericValue <= min;
  const incDisabled = numericValue >= max;

  return (
    <div>
      <label
        className="mb-4 block cursor-pointer text-xl text-primary sm:text-2xl"
        htmlFor={inputId}
      >
        {labelText}
      </label>
      <div className="flex">
        <button
          aria-label="decrement"
          type="button"
          className="h-16 w-16 cursor-pointer p-2 disabled:cursor-not-allowed disabled:opacity-20"
          disabled={decDisabled}
          {...getDecrementProps()}
        >
          <svg
            height="100%"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            fill="currentColor"
            viewBox="0 0 40 40"
            preserveAspectRatio="xMidYMid meet"
          >
            <g>
              <path d="m22.5 17.5v-5h-5v5h-5l7.5 10 7.5-10h-5z" />
            </g>
          </svg>
        </button>
        <input
          className="w-16 border border-solid border-current py-1 text-center text-2xl"
          id={inputId}
          pattern="[0-9]*"
          autoCapitalize="off"
          autoComplete="off"
          autoCorrect="off"
          spellCheck={false}
          {...getInputProps({
            onFocus: (e) => {
              e.target.select();
            },
          })}
        />
        <button
          aria-label="increment"
          type="button"
          className="h-16 w-16 cursor-pointer p-2 disabled:cursor-not-allowed disabled:opacity-20"
          disabled={incDisabled}
          {...getIncrementProps()}
        >
          <svg
            height="100%"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            fill="currentColor"
            viewBox="0 0 40 40"
            preserveAspectRatio="xMidYMid meet"
          >
            <g>
              <path d="m20 12.5l-7.5 10h5v5h5v-5h5l-7.5-10z" />
            </g>
          </svg>
        </button>
      </div>
    </div>
  );
}

CountStepper.propTypes = {
  inputId: PropTypes.string.isRequired,
  labelText: PropTypes.string,
  min: PropTypes.number,
  max: PropTypes.number,
  defaultValue: PropTypes.number,
  onNewValue: PropTypes.func,
};

CountStepper.defaultProps = {
  labelText: 'Count:',
  min: 1,
  max: 9,
  defaultValue: 1,
  onNewValue: () => {},
};

CountStepper.displayName = 'CountStepper';

export default CountStepper;
