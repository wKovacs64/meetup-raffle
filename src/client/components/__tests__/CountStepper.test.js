import React from 'react';
import { screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import { render } from '../../../test/utils';
import CountStepper from '../CountStepper';

const countStepperProps = {
  defaultValue: 5,
  inputId: 'foo',
  labelText: 'Count:',
};

describe('CountStepper', () => {
  it('increments', () => {
    render(<CountStepper {...countStepperProps} />);
    const input = screen.getByLabelText(countStepperProps.labelText);
    const initialIntValue = parseInt(input.value, 10);

    user.click(screen.getByRole('button', { name: /increment/i }));

    expect(parseInt(input.value, 10)).toBe(initialIntValue + 1);
  });

  it('decrements', () => {
    render(<CountStepper {...countStepperProps} />);
    const input = screen.getByLabelText(countStepperProps.labelText);
    const initialIntValue = parseInt(input.value, 10);

    user.click(screen.getByRole('button', { name: /decrement/i }));

    expect(parseInt(input.value, 10)).toBe(initialIntValue - 1);
  });

  it('accepts manual user input', async () => {
    render(<CountStepper {...countStepperProps} />);
    const input = screen.getByLabelText(countStepperProps.labelText);
    expect(input.value).not.toEqual('3');

    user.click(input);
    await user.type(input, '3');
    user.tab();

    expect(input.value).toBe('3');
  });

  it('coerces invalid input to the default value', async () => {
    render(<CountStepper {...countStepperProps} />);
    const input = screen.getByLabelText(countStepperProps.labelText);

    user.click(input);
    await user.type(input, '-');
    user.tab();

    expect(parseInt(input.value, 10)).toBe(countStepperProps.defaultValue);
  });

  it('maintains current value on blur when no changes were made', () => {
    render(<CountStepper {...countStepperProps} />);
    const input = screen.getByLabelText(countStepperProps.labelText);
    const initialValue = input.value;

    user.click(input);
    user.tab();

    expect(input.value).toBe(initialValue);
  });
});
