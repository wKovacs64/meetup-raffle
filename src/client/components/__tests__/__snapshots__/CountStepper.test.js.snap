// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountStepper renders 1`] = `
<div>
  <div>
    <div
      class="mb3"
    >
      <label
        class="f4 f3-ns lh-copy dark-blue"
        for="foo"
      >
        Count:
      </label>
    </div>
    <span
      class="flex flex-row"
    >
      <button
        aria-label="decrement"
        class="bn bg-transparent h3 w3 pointer near-black"
        data-testid="decrement-button"
        type="button"
      >
        <svg
          fill="currentColor"
          preserveAspectRatio="xMidYMid meet"
          viewBox="0 0 40 40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="m22.5 17.5v-5h-5v5h-5l7.5 10 7.5-10h-5z"
            />
          </g>
        </svg>
      </button>
      <input
        class="tc near-black w3 f3 pv1"
        id="foo"
        name="foo"
        pattern="[0-9]*"
        type="text"
        value="5"
      />
      <button
        aria-label="increment"
        class="bn bg-transparent h3 w3 pointer near-black"
        data-testid="increment-button"
        type="button"
      >
        <svg
          fill="currentColor"
          preserveAspectRatio="xMidYMid meet"
          viewBox="0 0 40 40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="m20 12.5l-7.5 10h5v5h5v-5h5l-7.5-10z"
            />
          </g>
        </svg>
      </button>
    </span>
  </div>
</div>
`;
